cadl-extension/src/code-model-builder.ts

```diff
@@ -1077,21 +1080,39 @@ export class CodeModelBuilder {
       return friendlyName;
     } else {
       if (target.kind === "Model" && target.templateArguments && target.templateArguments.length > 0) {
-        return (
-          target.name +
-          target.templateArguments
-            .map((it) => {
-              switch (it.kind) {
-                case "Model":
-                  return it.name;
-                case "String":
-                  return it.value;
-                default:
-                  return "";
-              }
-            })
-            .join("")
-        );
+        // hack for cadl-azure-core ResourceCreateOrUpdateModel or ResourceCreateOrReplaceModel
+        const knownCoreTemplate = new Set<string>([
+          "OptionalProperties",
+          "UpdateableProperties",
+          "DefaultKeyVisibility",
+        ]);
+
+        let modelUsedInCoreRequest = false;
+        while (knownCoreTemplate.has(target.name) && target.kind === "Model" && target.templateArguments && target.templateArguments.length > 0 && target.templateArguments[0].kind === "Model") {
+          modelUsedInCoreRequest = true;
+          target = target.templateArguments[0];
+        }
+
+        if (modelUsedInCoreRequest) {
+          return target.name + "Request";
```

Comment:
A tricky hack to make `OptionalProperties<UpdateableProperties<DefaultKeyVisibility<Project, "read">>>` result in a model named `ProjectRequest`.